lib/congress_forms.rb

```diff
@@ -1,21 +1,139 @@
 require "rest_client"
 
 module CongressForms
-  def self.url(path = "/", params = {}, bioguide_id = nil)
-    url = Rails.application.config.congress_forms_url + path
-    url += bioguide_id unless bioguide_id.nil?
-    url += "?" + {
+  class Form
+    attr_accessor :fields, :bioguide_id
+
+    def self.find(bioguide_ids)
+      raw_forms = CongressForms.post("/retrieve-form-elements/", { bio_ids: bioguide_ids })
+      raise CongressForms::RequestFailed if raw_forms.empty?
+      raw_forms.map { |id, raw| Form.new(id, raw["required_actions"]) }
+    end
+
+    def initialize(bioguide_id, fields)
+      @bioguide_id = bioguide_id
+      @fields = fields.map { |f| Field.new(f) }
+      order_fields
+    end
+
+    def order_fields
+      order = %w($NAME_PREFIX $NAME_FIRST $NAME_LAST $PHONE $EMAIL $SUBJECT $TOPIC)
+      @fields = @fields.sort_by { |f| order.index(f.value) || Float::INFINITY }
+    end
+
+    def fill(input, campaign_tag)
+      params = { bio_id: @bioguide_id, campaign_tag: campaign_tag, fields: input }
+      CongressForms.post("/fill-out-form/", params)
+    end
+  end
+
+  class Field
+    attr_accessor :max_length, :value
+    attr_writer :options_hash
+
+    def initialize(opts)
+      @max_length = opts["max_length"]
+      @value = opts["value"]
+      @options_hash = opts["options_hash"]
+    end
+
+    def validate(input)
+      return false if input.nil?
+      return false if max_length && input.length > max_length
+      return false unless options.nil? || options.include?(input)
+      true
+    end
+
+    def label
+      I18n.t value, scope: :congress_forms, default: value.sub("$", "").humanize
+    end
+
+    def is_select?
+      options_hash != nil
+    end
+
+    def ==(other)
+      max_length == other.max_length &&
+        value == other.value &&
+        options_hash == other.options_hash
+    end
+
+    def eql?(other)
+      self == other
+    end
+
+    def hash
+      [max_length, value, options_hash].hash
+    end
+
+    def options_hash
+      if @value == "$ADDRESS_STATE_POSTAL_ABBREV"
```

Comment:
Since this just returns a value based on a string comparison, I'm not sure the performance improvement would be worthwhile. I think we could add that behavior if setting up the options_hash becomes more complex.